database/plugin/metadata/sqlite/transaction.go

```diff
@@ -288,6 +288,130 @@ func (d *MetadataStoreSqlite) SetTransaction(
 			return result.Error
 		}
 	}
+	// Extract and save witness set data
+	// Delete existing witness records to ensure idempotency on retry
+	if tmpTx.ID != 0 {
+		if result := txn.Where("transaction_id = ?", tmpTx.ID).Delete(&models.KeyWitness{}); result.Error != nil {
+			return fmt.Errorf("delete existing key witnesses: %w", result.Error)
+		}
+		if result := txn.Where("transaction_id = ?", tmpTx.ID).Delete(&models.Script{}); result.Error != nil {
+			return fmt.Errorf("delete existing scripts: %w", result.Error)
+		}
+		if result := txn.Where("transaction_id = ?", tmpTx.ID).Delete(&models.Redeemer{}); result.Error != nil {
+			return fmt.Errorf("delete existing redeemers: %w", result.Error)
+		}
+		if result := txn.Where("transaction_id = ?", tmpTx.ID).Delete(&models.PlutusData{}); result.Error != nil {
+			return fmt.Errorf("delete existing plutus data: %w", result.Error)
+		}
+	}
+	if tx.Witnesses() != nil {
+		ws := tx.Witnesses()
+
+		// Add Vkey Witnesses
+		for _, vkey := range ws.Vkey() {
+			keyWitness := models.KeyWitness{
+				TransactionID: tmpTx.ID,
+				Type:          0, // VkeyWitness
```

Comment:
We should define our own constants for the "key witness" type, since there aren't any provided by gOuroboros.